### PR TITLE
Add `BoundedBTreeSet`

### PR DESCRIPTION
**Pull request by coriolinus**
Part of https://github.com/paritytech/substrate/issues/8719

### CODE REVIEW

**Comment by bkchr** [comment]
IMHO we have now reached the point where this should be abstracted. IMHO I don't see any point in copying this for every type in the std library (I know that we don't need every type). However this is mainly code copying. Which should be doable with some generic container or a macro that generates the stuff.

**Comment by coriolinus** [comment]
I'm not sure how well it can be automated. The desired abstraction involves

- Deref / AsRef impls for appropriate types
- passthrough methods for `&mut self` methods which cannot increase the size
- `try_X` methods for `&mut self` methods which can increase the size

Knowing which types are appropriate, and which methods can increase the size, seems to require human intervention.

That said, once this is in, we'll have bounded variants for lists, sets, and maps. Those should suffice for all reasonable storage cases. I'm not anticipating adding more PRs like this soon.

File: frame/support/src/storage/bounded_btree_set.rs [hunk 1]
```diff
@@ -0,0 +1,408 @@
...
+	/// let bounded_map = unsafe {BoundedBTreeSet::<_, Size>::force_from(set, "demo")};
+	/// ```
+	pub unsafe fn force_from<Scope>(set: BTreeSet<T>, scope: Scope) -> Self
+	where
+		Scope: Into<Option<&'static str>>,
+	{
+		if set.len() > Self::bound() {
+			log::warn!(
+				target: crate::LOG_TARGET,
+				"length of a bounded btreemap in scope {} is not respected.",
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_set.rs:92]:
```suggestion
				"length of a bounded btreeset in scope {} is not respected.",
```

File: frame/support/src/storage/bounded_btree_set.rs [hunk 2]
```diff
@@ -0,0 +1,408 @@
...
+	/// ```
+	/// # use sp_std::collections::btree_set::BTreeSet;
+	/// # use frame_support::{parameter_types, storage::bounded_btree_set::BoundedBTreeSet};
+	/// parameter_types! {
+	/// 	pub const Size: u32 = 5;
+	/// }
+	/// let mut set = BTreeSet::new();
+	/// set.insert("foo");
+	/// set.insert("bar");
+	/// let bounded_map = unsafe {BoundedBTreeSet::<_, Size>::force_from(set, "demo")};
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_set.rs:83]:
```suggestion
	/// let bounded_set = unsafe {BoundedBTreeSet::<_, Size>::force_from(set, "demo")};
```

File: frame/support/src/storage/bounded_btree_set.rs [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+		<BTreeSet<T> as codec::DecodeLength>::len(self_encoded)
+	}
+}
+
+impl<T, S> StorageDecodeLength for BoundedBTreeSet<T, S> {}
+
+impl<T, S> codec::EncodeLike<BTreeSet<T>> for BoundedBTreeSet<T, S> where
+	BTreeSet<T>: Encode
+{
+}
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_set.rs:306]:
```suggestion
{}
```

File: frame/support/src/storage/bounded_btree_set.rs [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+			self.0.insert(item);
+			Ok(())
+		} else {
+			Err(())
+		}
+	}
+
+	/// Remove an item from the set, returning whether it was previously in the set.
+	///
+	/// The item may be any borrowed form of the set's item type, but the ordering on the borrowed
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_set.rs:139]: TIL 👍

File: frame/support/src/storage/bounded_btree_set.rs [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+impl<T, S> AsRef<BTreeSet<T>> for BoundedBTreeSet<T, S>
+where
+	T: Ord,
+{
+	fn as_ref(&self) -> &BTreeSet<T> {
+		&self.0
+	}
+}
+
+impl<T, S> From<BoundedBTreeSet<T, S>> for BTreeSet<T>
```
Comment (kianenigma) [inline frame/support/src/storage/bounded_btree_set.rs:271]: I think we might miss this in `BoundedVec` and `BoundedBTreeMap`, good one!

**Review by kianenigma — APPROVED** [review]

**Comment by kianenigma** [comment]
> IMHO we have now reached the point where this should be abstracted. IMHO I don't see any point in copying this for every type in the std library (I know that we don't need every type). However this is mainly code copying. Which should be doable with some generic container or a macro that generates the stuff.

I don't have as strong opinion about BtreeMap and Set. On one hand they are types that we support in storage and we are a library, on the other hand we don't really use them (for _good_ reasons) and we shouldn't overkill it. Either way I am happy with this PR. 

About your comment on abstraction: I think what we do need is a `SortedVec` next to our `BoundedVec`, because we have quite a few storage items that have the assumption that they must always be sorted. And implementation-wise I was thinking of something like this: 

```rust 
struct AttributeVec<T, Attr: Attribute>(Vec<T>);

/// An attribute to a container like a Vec
trait Attribute {
  fn pre_insert();
  fn post_insert();
  fn pre_delete(); 
  fn post_delete(); 
  // and some other hooks/utilities that allow an attribute to control how a container mutates.
}

/// An attribute that ensures a container's length never goes beyond a limit.
struct Bounded;
impl Attribute for Bounded {
  fn pre_insert() { /* ensure bound */}
}

/// An attribute that ensures a container is always sorted. 
struct Sorted;
impl Attribute for Sorted {
  fn post_insert() { /* do sort! */}
  fn post_delete() { /* do sort! */}
}

// and we can have other attributes, and attributes can be combined by aggregating them into a tuple. 
#[impl_trait_for_tuples(18)]
if Attribute for Tuple { /* ... */ }

/// 
type BoundedVec<T> = AttributeVec<T, Bounded>;
type SortedVec<T> = AttributeVEc<T, Sorted>;
type BoundedSortedVec<T> = AttributeVEc<T, (Bounded, Sorted)>;
```

But to be honest there are a lot of empty pieces in my brain dump above, and I am worried that this would add so much complexity that we would regret it and just go back to the simpler way of copy-pasta.

**Comment by joao-paulo-parity** [comment]
Hello?

**Comment by kianenigma** [comment]
do you love me cla bot?

**Review by shawntabrizi — APPROVED** [review]

**Comment by coriolinus** [comment]
bot merge

**Pull request merged**